File: archive/benchmark/bench.cpp
#include <array>
#include <chrono>
#include <cstdint>
#include <cstring>
#include <iomanip>
#include <iostream>

static inline uint64_t xorshift64(uint64_t &s) {
    s ^= s << 13;
    s ^= s >> 7;
    s ^= s << 17;
    return s;
}

int main(int argc, char **argv) {
    uint64_t steps = 50000000ULL;
    if (argc > 1) steps = std::strtoull(argv[1], nullptr, 10);

    std::array<uint8_t, 65536> mem{};
    uint8_t a = 0x12, x = 0x34, y = 0x56, p = 0x24;
    uint16_t pc = 0x4000;
    uint64_t state = 0x65022026ULL;

    for (size_t i = 0; i < mem.size(); ++i) mem[i] = static_cast<uint8_t>(i ^ (i >> 8));

    auto t0 = std::chrono::high_resolution_clock::now();
    for (uint64_t i = 0; i < steps; ++i) {
        uint8_t op = static_cast<uint8_t>(xorshift64(state) & 0xFF);
        uint16_t addr = static_cast<uint16_t>(pc + x + (y << 1));
        uint8_t m = mem[addr];

        switch (op & 0x0F) {
            case 0:
                a = static_cast<uint8_t>(a + m + (p & 1));
                p = (p & ~0x83) | (a == 0 ? 0x02 : 0) | (a & 0x80 ? 0x80 : 0);
                break;
            case 1:
                a ^= m;
                p = (p & ~0x82) | (a == 0 ? 0x02 : 0) | (a & 0x80 ? 0x80 : 0);
                break;
            case 2:
                a |= m;
                p = (p & ~0x82) | (a == 0 ? 0x02 : 0) | (a & 0x80 ? 0x80 : 0);
                break;
            case 3:
                a &= m;
                p = (p & ~0x82) | (a == 0 ? 0x02 : 0) | (a & 0x80 ? 0x80 : 0);
                break;
            case 4:
                mem[addr] = static_cast<uint8_t>(m + x);
                break;
            case 5:
                mem[addr] = static_cast<uint8_t>(m - y);
                break;
            case 6:
                x = static_cast<uint8_t>(x + 1);
                break;
            case 7:
                y = static_cast<uint8_t>(y - 1);
                break;
            case 8:
                pc = static_cast<uint16_t>((static_cast<int32_t>(pc) + static_cast<int8_t>(m)) & 0xFFFF);
                break;
            case 9:
                p ^= 0x41;
                break;
            case 10:
                a = static_cast<uint8_t>((a << 1) | (a >> 7));
                break;
            case 11:
                a = static_cast<uint8_t>((a >> 1) | (a << 7));
                break;
            case 12:
                mem[static_cast<uint16_t>(pc + x)] ^= a;
                break;
            case 13:
                mem[static_cast<uint16_t>(pc + y)] += p;
                break;
            case 14:
                x ^= y;
                break;
            default:
                y += a;
                break;
        }
        pc = static_cast<uint16_t>(pc + 1);
    }
    auto t1 = std::chrono::high_resolution_clock::now();
    double secs = std::chrono::duration<double>(t1 - t0).count();
    double mops = (steps / 1e6) / secs;

    uint64_t checksum = static_cast<uint64_t>(a) | (static_cast<uint64_t>(x) << 8) |
                        (static_cast<uint64_t>(y) << 16) | (static_cast<uint64_t>(p) << 24) |
                        (static_cast<uint64_t>(pc) << 32) | (static_cast<uint64_t>(mem[0x1234]) << 48);

    std::cout << "lang=cpp steps=" << steps << " seconds=" << std::fixed << std::setprecision(6)
              << secs << " mops=" << std::setprecision(3) << mops << " checksum=0x" << std::hex
              << checksum << std::dec << "\n";
    return 0;
}

File: archive/benchmark/bench.rs
use std::env;
use std::time::Instant;

#[inline(always)]
fn xorshift64(mut s: u64) -> u64 {
    s ^= s << 13;
    s ^= s >> 7;
    s ^= s << 17;
    s
}

fn main() {
    let steps: u64 = env::args()
        .nth(1)
        .and_then(|s| s.parse::<u64>().ok())
        .unwrap_or(50_000_000);

    let mut mem = [0u8; 65536];
    let mut a: u8 = 0x12;
    let mut x: u8 = 0x34;
    let mut y: u8 = 0x56;
    let mut p: u8 = 0x24;
    let mut pc: u16 = 0x4000;
    let mut state: u64 = 0x6502_2026;

    for i in 0..mem.len() {
        mem[i] = (i as u8) ^ ((i >> 8) as u8);
    }

    let t0 = Instant::now();
    for _ in 0..steps {
        state = xorshift64(state);
        let op = (state & 0xFF) as u8;
        let addr = pc.wrapping_add(x as u16).wrapping_add((y as u16) << 1);
        let m = mem[addr as usize];

        match op & 0x0F {
            0 => {
                a = a.wrapping_add(m).wrapping_add(p & 1);
                p = (p & !0x83) | if a == 0 { 0x02 } else { 0 } | if (a & 0x80) != 0 { 0x80 } else { 0 };
            }
            1 => {
                a ^= m;
                p = (p & !0x82) | if a == 0 { 0x02 } else { 0 } | if (a & 0x80) != 0 { 0x80 } else { 0 };
            }
            2 => {
                a |= m;
                p = (p & !0x82) | if a == 0 { 0x02 } else { 0 } | if (a & 0x80) != 0 { 0x80 } else { 0 };
            }
            3 => {
                a &= m;
                p = (p & !0x82) | if a == 0 { 0x02 } else { 0 } | if (a & 0x80) != 0 { 0x80 } else { 0 };
            }
            4 => mem[addr as usize] = m.wrapping_add(x),
            5 => mem[addr as usize] = m.wrapping_sub(y),
            6 => x = x.wrapping_add(1),
            7 => y = y.wrapping_sub(1),
            8 => pc = pc.wrapping_add(m as i8 as i16 as u16),
            9 => p ^= 0x41,
            10 => a = a.rotate_left(1),
            11 => a = a.rotate_right(1),
            12 => {
                let ix = pc.wrapping_add(x as u16) as usize;
                mem[ix] ^= a;
            }
            13 => {
                let iy = pc.wrapping_add(y as u16) as usize;
                mem[iy] = mem[iy].wrapping_add(p);
            }
            14 => x ^= y,
            _ => y = y.wrapping_add(a),
        }
        pc = pc.wrapping_add(1);
    }

    let secs = t0.elapsed().as_secs_f64();
    let mops = (steps as f64 / 1_000_000.0) / secs;
    let checksum: u64 = (a as u64)
        | ((x as u64) << 8)
        | ((y as u64) << 16)
        | ((p as u64) << 24)
        | ((pc as u64) << 32)
        | ((mem[0x1234] as u64) << 48);

    println!(
        "lang=rust steps={} seconds={:.6} mops={:.3} checksum=0x{:x}",
        steps, secs, mops, checksum
    );
}
